Release: Create pre release script

spinscale opened the pull request:
Refactored a part out of the release script, so the user can
change the version locally as well as move the documentation
and change the Version.java

The background of this change is to have a very simple release
process that puts stuff into a staging environment, so the beta
release can be tested, before it is officially released.

This means the build_release script can be removed soon.

spinscale commented:
@s1monw would be awesome if you could take a look, but I basically copied most of the existing release..

you can easily run the script in your local branch, as it does not push, it just adds two commits to your local branch, which can easily be reverted again

dadoonet commented on core/src/main/java/org/elasticsearch/plugins/PluginManager.java:
```diff
@@ -430,9 +430,7 @@ public void listInstalledPlugins(Terminal terminal) throws IOException {
-                    if (Version.CURRENT.snapshot()) {
-                        addUrl(urls, String.format(Locale.ROOT, "http://download.elastic.co/elasticsearch/snapshot/org/elasticsearch/plugin/%s/%s-SNAPSHOT/%s-%s-SNAPSHOT.zip", repo, version, repo, version));
-                    }
+                    addUrl(urls, String.format(Locale.ROOT, "http://download.elastic.co/elasticsearch/staging/org/elasticsearch/plugin/%s/%s/%s-%s.zip", repo, version, repo, version));
```
just wondering if we should not invert the two lines. 
Most of the time, we will have only to check the first URL.

spinscale replied:
makes total sense, will fix

s1monw commented on dev-tools/prepare_release.py:
```diff
@@ -0,0 +1,165 @@
+      if process_file(full_path, callback):
+        pending_files.append(os.path.join(root, file_name))
+  return pending_files
+
```
usually in python you do this in a _main_ method / block like this:

```python
if __name__ == "__main__":
  release_version = find_release_version()
  print('*** Preparing release version: [%s]' % release_version)
  ...
```
that way you can also import the file and use it's methods without executing it

spinscale replied:
done

s1monw commented on dev-tools/prepare_release.py:
```diff
@@ -0,0 +1,165 @@
+# an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+# either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
```
can we summarize what this script does? it would we awesome to have an example how to run it too :)

spinscale replied:
done :)

s1monw commented on core/src/main/java/org/elasticsearch/plugins/PluginManager.java:
```diff
@@ -430,10 +430,8 @@ public void listInstalledPlugins(Terminal terminal) throws IOException {
-                        addUrl(urls, String.format(Locale.ROOT, "http://download.elastic.co/elasticsearch/snapshot/org/elasticsearch/plugin/%s/%s-SNAPSHOT/%s-%s-SNAPSHOT.zip", repo, version, repo, version));
-                    }
                     addUrl(urls, String.format(Locale.ROOT, "http://download.elastic.co/elasticsearch/release/org/elasticsearch/plugin/%s/%s/%s-%s.zip", repo, version, repo, version));
+                    addUrl(urls, String.format(Locale.ROOT, "http://download.elastic.co/elasticsearch/staging/org/elasticsearch/plugin/%s/%s/%s-%s.zip", repo, version, repo, version));
```
I have been thinking about this and I think we should enable this with a flag and not by default. Can we tell the download manager via some setting to look at the staging path? Another option would be to deploy into `release` as we did before and keep that part. If we have to re-run a release we can still just redeploy to S3. I mean we should try this out if it works to override the existing deployed artifacts but I think it should? This would simplify the process on S3 again and we would only need to flip the switch on sonatype?

spinscale replied:
philosophical question I guess... as there is no staging with S3 like on sonatype, I would prefer to not be able to push into the final repo directly (and esepcially accidentally)... coming up with a flag solution

dadoonet replied:
Yes. A flag looks a good solution to me.
We could also autodetect SNAPSHOT in the provided String and add all SNAPSHOT repos in that case.

Can you add Sonatype SNAPSHOT repo BTW?

s1monw commented:
I left some commetns - looks good though!

s1monw commented on dev-tools/prepare_release.py:
```diff
@@ -0,0 +1,181 @@
+
+# Executes a git commit with 'release [version]' as the commit message
+def commit_release(release):
+  run('git commit -m "release [%s]"' % release)
```
can we just call this `Flip version from x to y" instead of `release [y]` it looks odd in the history

spinscale replied:
done

s1monw commented on dev-tools/prepare_release.py:
```diff
@@ -0,0 +1,181 @@
+# Moves the Version.java file from a snapshot to a release
+def remove_version_snapshot(version_file, release):
+  # 1.0.0.Beta1 -> 1_0_0_Beta1
+  release = release.replace('.', '_')
```
here we also need to s/-/_/ for beta1 since it's `2.0.0-beta1` but needs to be turned into `2_0_0_beta1`

s1monw replied:
just add `release = release.replace('-', '_')` I guess?

spinscale replied:
yeah, fixed

s1monw commented on dev-tools/prepare_release.py:
```diff
@@ -0,0 +1,181 @@
+  replacement = 'new Version(V_%s_ID, false' % (release)
+  def callback(line):
+    return line.replace(pattern, replacement)
+  process_file(version_file, callback)
```
I also think we should validate this like this:

```python
 processed = process_file(version_file, callback)
  if not processed:
    raise RuntimeError('failed to remove snapshot version for %s' % (release))
```

that way we should have got an exception

s1monw replied:
btw. I tested this ^^ :)

spinscale replied:
added. thx for testing :)

s1monw commented on dev-tools/prepare_release.py:
```diff
@@ -0,0 +1,182 @@
+
+  pending_files = update_reference_docs(release_version)
+  # split commits for docs and version to enable easy cherry-picking
+  if pending_files:
```
can we add add

```python
  else:
    print(`WARNING: no documentation references updates for release %s` % (release_version))
```

spinscale replied:
added

s1monw commented:
LGTM

spinscale committed:
Release: Create pre release script

Refactored a part out of the release script, so the user can
change the version locally as well as move the documentation
and change the Version.java

The background of this change is to have a very simple release
process that puts stuff into a staging environment, so the beta
release can be tested, before it is officially released.

This means the build_release script can be removed soon.
